util/UnitEvent.lua

```diff
@@ -0,0 +1,186 @@
+--[[
+    UnitEvent (Namesake)
+        Originally by Bribe
+
+    Uses Magic Defense instead of Defend ability to detect leave.    
+]]
+
+require('CeresStdLib.base.Rawcode')
+require('CeresStdLib.base.Optimizations')
+require('CeresStdLib.base.Native')
+require('CeresStdLib.base.Init')
+require('CeresStdLib.base.Log')
+
+require('CeresStdLib.util.World')
+
+DETECT_LEAVE            = fromRawCode('Amdf')
+
+EVENT_UNIT_ENTER        = 1
+EVENT_UNIT_LEAVE        = 2
+
+UnitEvent = {}
+
+local trigTable         = {}
+local __userData        = {}
+__userData[-1]          = {}   --  Holds references to each unit
+__userData[-2]          = {}   --  Holds preplaced flag
+__userData[0]           = 0
+
+trigTable[3], trigTable[4]                  = {}, {}
+
+local __evType          = 0
+local __evData          = {0, nil}
+local __initialized     = false
+
+replaceNative('GetUnitUserData', function(u)
+    return __userData[u]

```

Comment:
I heard handles in lua fail to be consistent keys in tables. Consider referencing them by id.

Reply:
So, by using GetHandleId? Will try to write that up.

Reply:
maybe, but maybe better with our own unique ID assigned with `Native.SetUnitUserData`.
Although it doesn't matter now since we don't have arrays anymore, only maps.